cylc/flow/host_select.py

```diff
@@ -0,0 +1,532 @@
+"""Functionality for selecting a host from pre-defined list."""
+import ast
+from collections import namedtuple
+from functools import lru_cache
+from io import BytesIO
+from itertools import dropwhile
+import json
+import random
+from time import sleep
+import token
+from tokenize import tokenize
+
+from cylc.flow import LOG
+from cylc.flow.cfgspec.glbl_cfg import glbl_cfg
+from cylc.flow.exceptions import HostSelectException
+from cylc.flow.hostuserutil import get_fqdn_by_host, is_remote_host
+from cylc.flow.remote import remote_cylc_cmd, run_cmd
+
+
+def select_suite_host(cached=True):
+    """Return a host as specified in `[suite hosts]`.
+
+    * Condemned hosts are filtered out.
+    * Filters by thresholds (if defined).
+    * Ranks by thresholds (if defined).
+
+    Args:
+        cached (bool):
+            Use a cached version of the global configuration if True
+            else reload from the filesystem.
+
+    Returns:
+        tuple - See `select_host` for details.
+
+    Raises:
+        HostSelectException: See `select_host` for details.
+
+    """
+    # get the global config, if cached = False a new config instance will
+    # be returned with the up-to-date configuration.
+    global_config = glbl_cfg(cached=cached)
+
+    return select_host(
+        # list of suite hosts
+        global_config.get(['suite servers', 'run hosts']) or ['localhost'],
+        # thresholds / ranking to apply
+        threshold_string=global_config.get(['suite servers', 'thresholds']),
+        # list of condemned hosts
+        blacklist=global_config.get(
+            ['suite servers', 'condemned hosts']
+        ),
+        blacklist_name='condemned host'
+    )
+
+
+def select_host(
+        hosts,
+        threshold_string=None,
+        blacklist=None,
+        blacklist_name=None
+):
+    """Select a host from the provided list.
+
+    If no ranking is provided (in `threshold_string`) then random selection
+    is used.
+
+    Args:
+        threshold_string (str):
+            A multiline string containing Python expressions to fileter
+            hosts by e.g::
+
+               # only consider hosts with less than 70% cpu usage
+               # and a server load of less than 5
+               cpu_percent() < 70
+               getloadavg()[0] < 5
+
+            And or Python statements to rank hosts by e.g::
+
+               # rank by used cpu, then by load average as a tie-break
+               # (lower scores are better)
+               cpu_percent()
+               getloadavg()
+
+            Comments are allowed using `#` but not inline comments.
+        blacklist (list):
+            List of host names to filter out.
+            Can be short host names (do not have to be fqdn values)
+        blacklist_name (str):
+            Used with `blacklist` in error messages.
```

Comment:
I found the description "The reason for blacklisting these hosts (used for exceptions)" further down the document more helpful than this.